### DISCUSSION TITLE
Include window functions in duckdb_functions()?

### DISCUSSION THREAD

**Discussion by vlulla** [duckdb/duckdb#14353] [General]
I was surprised to see that the query 
```sql
select * from duckdb_functions() where function_name='lead';
```
yielded no results. Therefore, I checked to see if i could get any results for `lag` (or a few other window functions). I was surprised that none of the window functions are included in `duckdb_functions()`! 

I am curious about these two items:
1. Why are window functions not included in `duckdb_functions()`?
1. Is it possible to get a list of window functions, and their parameter argument and return types, in the cli? 

When i'm exploring/manipulating the data, i often consult `duckdb_functions()` to quickly glance at the various parameters, and their corresponding types, that functions accept. For this instance, i was reading about [AsOf joins](https://duckdb.org/2023/09/15/asof-joins-fuzzy-temporal-lookups) and it required me to lookup `lead` function's function signature (particularly for the `default` parameter argument). So, i tried searching for the function definition in my cli session and i was disappointed that i had to go to the webpage of windows function to learn about the arguments of `lead`!

**Comment by Tishj** [comment]
That's a good point, it's likely due to the windowing functions not being bound through the same process as functions/aggregates
Something that should probably be fixed/changed at some point

**Reply by vlulla** [reply]
I read. Thanks for the explanation @Tishj . I find reflection (for e.g., system catalog in PostgreSQL) indispensable for my exploratory work.

**Comment by hawkfish** [comment]
@vlulla What @Tishj said is exactly right - most window functions are parsed separately. The exception is that all aggregate functions can be used in a windowing context. 

Which begs the question as to whether we should list them twice? Plus there are restrictions on windowed aggregates (like no secondary `ORDER BY` clauses e.g., `FIRST(x ORDER BY y) OVER (ORDER BY z)`), so it could be misleading to list them separately as window functions?  I'm in the process of removing some of those restrictions, but I'm not sure there are techniques for all of them.

**Reply by vlulla** [reply]
Hmm. I understand the tentativeness for listing functions twice. Why not list _all_ functions at least once? And, why can't uniqueness be based on the function type (i.e., combination of parameter and return types) and not on whether the function is in an aggregation/windowing context? Regardless, it appears that _not listing_ window functions _even once_ appears unwise.

Re your point about secondary `ORDER BY` clauses: I have never used, or even thought, of this! I'm curious whether PostgreSQL also has this restriction.

Regardless, thanks @hawkfish for your thoughts.

**Comment by hawkfish** [comment]
This was added to 1.5, but for 2.0 all window functions are in the catalog, so even new ones in extensions will be visible.

**Reply by vlulla** [reply]
Great to read! Thank you @hawkfish !
